bftengine/src/bftengine/ReplicaImp.hpp

```diff
@@ -457,30 +459,14 @@ class ReplicaImp : public InternalReplicaApi, public IReplicaForStateTransfer {
 
   virtual const ReplicasInfo& getReplicasInfo() override { return (*repsInfo); }
 
-  virtual Timer& getViewChangeTimer() override { return *viewChangeTimer; }
-
-  virtual Timer& getStateTranTimer() override { return *stateTranTimer; }
-
-  virtual Timer& getRetransmissionsTimer() override { return *retranTimer; }
-
-  virtual Timer& getStatusTimer() override { return *statusReportTimer; }
-
-  virtual Timer& getSlowPathTimer() override { return *slowPathTimer; }
-
-  virtual Timer& getInfoRequestTimer() override { return *infoReqTimer; }
-
-  virtual Timer& getDebugStatTimer() override { return *debugStatTimer; }
-
-  virtual Timer& getMetricsTimer() override { return *metricsTimer_; }
-
-  virtual void onViewsChangeTimer(Time cTime, Timer& timer) override;
-  virtual void onStateTranTimer(Time cTime, Timer& timer) override;
-  virtual void onRetransmissionsTimer(Time cTime, Timer& timer) override;
-  virtual void onStatusReportTimer(Time cTime, Timer& timer) override;
-  virtual void onSlowPathTimer(Time cTime, Timer& timer) override;
-  virtual void onInfoRequestTimer(Time cTime, Timer& timer) override;
-  virtual void onDebugStatTimer(Time cTime, Timer& timer) override;
-  virtual void onMetricsTimer(Time cTime, Timer& timer) override;
+  virtual void onViewsChangeTimer(concordUtil::Timers::Handle);
+  virtual void onStateTranTimer(concordUtil::Timers::Handle);
+  virtual void onRetransmissionsTimer(concordUtil::Timers::Handle);
+  virtual void onStatusReportTimer(concordUtil::Timers::Handle);
+  virtual void onSlowPathTimer(concordUtil::Timers::Handle);
+  virtual void onInfoRequestTimer(concordUtil::Timers::Handle);
+  virtual void onDebugStatTimer(concordUtil::Timers::Handle);
+  virtual void onMetricsTimer(concordUtil::Timers::Handle);
```

Comment:
Based on `ReplicaImp`'s interface and implementation, I think these can be made non-virtual. For example, an inheriting class is not able to change message handling and I would assume that is also the case for timer handling.

Reply:
Good catch.